### PR TITLE
Add ObjectDisposedException.Throw for object instance and type

### LINKED ISSUE #51700: Add new ObjectDisposedException constructor overload
## Background and Motivation

There is a very common pattern used to handle disposed object state and it involves throwing `ObjectDisposedException`. It requires extra conversion before constructing the exception. There are about 400 cases of this pattern in runtime which could be unified and simplified and more in .NET libraries.

```c#
	throw new ObjectDisposedException(GetType().FullName);
	// or
	throw new ObjectDisposedException(GetType().Name);
	// or
	throw new ObjectDisposedException(GetType().ToString());
	// or
	throw new ObjectDisposedException(nameof(NameOfThisType));
```

## Proposed API

```diff
namespace System
{
     public class ObjectDisposedException {
+		public ObjectDisposedException(object instance);
     }
```

## Usage Examples

All the examples from above could be replaced with simple construction which would extract the name from the instance.

```c#
	throw new ObjectDisposedException(this);
```

### PR DESCRIPTION
**Pull request by Bibletoon**
Closes #51700
I've decided to make [static methods](https://github.com/dotnet/runtime/issues/51700#issuecomment-879345427) instead of new ctors because of breaking changes with null argument (like [there](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Threading/ReaderWriterLockSlim.cs#L298) and [there](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Security.Cryptography.Primitives/src/System/Security/Cryptography/HashAlgorithm.cs#L36))

Should I also change old calls like that ```throw new ObjectDisposedException(GetType().FullName);``` or ```throw new ObjectDisposedException(GetType().Name);``` to new method?

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=58684) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=58684) before we can accept your contribution.<br/><br/>:x: Bibletoon [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=58684)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=58684) it.</sub>

File: src/libraries/System.Private.CoreLib/src/System/ObjectDisposedException.cs
Comment (teo-tsirpanis) [inline src/libraries/System.Private.CoreLib/src/System/ObjectDisposedException.cs:48]: Should they also have `[StackTraceHidden]`? Otherwise, the first `Throw` method should not call the second, for a cleaner stacktrace.

File: src/libraries/System.Private.CoreLib/src/System/ObjectDisposedException.cs
Comment (martincostello) [inline src/libraries/System.Private.CoreLib/src/System/ObjectDisposedException.cs:50]:
```suggestion
            Throw(instance?.GetType());
```

File: src/libraries/System.Private.CoreLib/src/System/ObjectDisposedException.cs
Comment (martincostello) [inline src/libraries/System.Private.CoreLib/src/System/ObjectDisposedException.cs:54]:
```suggestion
            throw new ObjectDisposedException(type?.FullName);
```

File: src/libraries/System.Private.CoreLib/src/System/ObjectDisposedException.cs
Comment (martincostello) [inline src/libraries/System.Private.CoreLib/src/System/ObjectDisposedException.cs:49]: As the existing constructors allow for null.

```suggestion
        public static void Throw(object? instance) =>
```

File: src/libraries/System.Private.CoreLib/src/System/ObjectDisposedException.cs
Comment (martincostello) [inline src/libraries/System.Private.CoreLib/src/System/ObjectDisposedException.cs:53]:
```suggestion
        public static void Throw(Type? type) =>
```

File: src/libraries/System.Private.CoreLib/src/System/ObjectDisposedException.cs
Comment (marek-safar) [inline src/libraries/System.Private.CoreLib/src/System/ObjectDisposedException.cs:55]: Should this be nullable?

**Comment by marek-safar** [comment]
> Should I also change old calls like that throw new ObjectDisposedException(GetType().FullName); or throw new ObjectDisposedException(GetType().Name); to new method?

Yes please but in separate PR.

File: src/libraries/System.Private.CoreLib/src/System/ObjectDisposedException.cs
Comment (martincostello) [reply src/libraries/System.Private.CoreLib/src/System/ObjectDisposedException.cs:55]: I suggested it was nullable as the existing constructors allow null, and it can't be relied upon for null checking as the consuming app might not have nullable enabled. Otherwise they'd get a `NullReferenceException` if the type/object passed to the new methods was null (without adding a null check guard to the new methods, which seems to go against the null-permissive existing constructos).

**Comment by stephentoub** [comment]
> I've decided to make static methods instead of new ctors because of breaking changes with null argument (like there and there)

The APIs were approved as constructors.  We can't merge this with different APIs unless that's officially revisited (my comment about using static members instead was just a musing).
cc: @terrajobst, @bartonjs

**Comment by stephentoub** [comment]
>> Should I also change old calls like that throw new ObjectDisposedException(GetType().FullName); or throw new ObjectDisposedException(GetType().Name); to new method?

> Yes please but in separate PR.

FWIW, my preference is actually to do it in the same PR.  That helps to validate the design of the API by seeing whether there are any places we'd like to change that can't be changed for some reason, or otherwise discovering issues with the shape of the API.  That's still possible if it's rolled out separately, but only after it's already been merged, at which point we then need to go back and revise the public API, which could have already been adopted elsewhere up-stack by that point, and now we need to deal with a breaking change (within the same release).

**Comment by tmds** [comment]
Maybe also add?
```cs
public static void Throw<T>() => Throw(typeof(T));
```

**Comment by bartonjs** [comment]
I've reset the issue to api-ready-for-review to discuss the implementation-vs-spec mismatch.

File: src/libraries/Common/tests/System/Net/WebSockets/WebSocketStream.cs
Comment (jkotas) [inline src/libraries/Common/tests/System/Net/WebSockets/WebSocketStream.cs:277]: This method should be deleted, and `ObjectDisposedException.Throw(this)` should be called directly instead.

File: src/libraries/System.Runtime/ref/System.Runtime.cs
Comment (jkotas) [inline src/libraries/System.Runtime/ref/System.Runtime.cs:4973]: `StackTraceHidden` should not be in the reference assembly. It should be just in the implementation.

**Comment by Bibletoon** [comment]
After all those failed checks, I guess I should remove DoesNotReturn attribute, am I right?

File: src/libraries/Common/tests/System/IO/PositionValueStream.cs
Comment (stephentoub) [inline src/libraries/Common/tests/System/IO/PositionValueStream.cs:23]:
```suggestion
                ObjectDisposedException.Throw(this);
```

File: src/libraries/System.Collections.Immutable/src/Validation/Requires.cs
Comment (stephentoub) [inline src/libraries/System.Collections.Immutable/src/Validation/Requires.cs:143]: This library builds for TFMs other than .NET 7.  You'll need to revert changes to code in any such libraries (it could be ifdef'd, but I don't think it's worth doing so).

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpContentStream.cs
Comment (stephentoub) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpContentStream.cs:44]: We should delete this method and change the one call site above to instead be:
```C#
HttpConnection c = _connection;
if (c is null)
{
    ObjectDisposedException.Throw(this);
}
return c;
```

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/NetworkStream.cs
Comment (stephentoub) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/NetworkStream.cs:677]: We should delete this helper and just change ThrowIfDisposed to be:
```C#
if (_disposed != 0)
{
    ObjectDisposedException.Throw(this);
}
```

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs
Comment (stephentoub) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:3819]: We should delete this helper and just change ThrowIfDisposed above to be:
```C#
if (Disposed)
{
    ObjectDisposedException.Throw(this);
}
```

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/TCPClient.cs
Comment (stephentoub) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/TCPClient.cs:461]: Ditto

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/TCPListener.cs
Comment (stephentoub) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/TCPListener.cs:301]:
```suggestion
                ObjectDisposedException.Throw(typeof(Socket));
                throw;
```

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/UDPClient.cs
Comment (stephentoub) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/UDPClient.cs:1098]: As earlier, delete the helper and use the new method in ThrowIfDisposed above.

File: src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/ClientWebSocket.cs
Comment (stephentoub) [inline src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/ClientWebSocket.cs:72]: Probably simplest just to revert this line rather than trying to work around the compiler not knowing that Throw never returns.

File: src/libraries/System.Private.CoreLib/src/System/ObjectDisposedException.cs
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/ObjectDisposedException.cs:51]: This was discussed and the conclusions was it should not be nullable:
https://github.com/dotnet/runtime/issues/51700#issuecomment-924242084

File: src/libraries/System.Private.CoreLib/src/System/ObjectDisposedException.cs
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/ObjectDisposedException.cs:56]: This was discussed and the conclusions was it should not be nullable:
https://github.com/dotnet/runtime/issues/51700#issuecomment-924242084

File: src/libraries/System.Private.CoreLib/src/System/ObjectDisposedException.cs
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/ObjectDisposedException.cs:52]: But even without the argument being nullable, I agree with @martincostello that if null is passed in, it shouldn't null ref.

File: src/libraries/System.Private.CoreLib/src/System/ObjectDisposedException.cs
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/ObjectDisposedException.cs:57]: Ditto

File: src/libraries/Common/tests/System/IO/PositionValueStream.cs
Comment (bartonjs) [reply src/libraries/Common/tests/System/IO/PositionValueStream.cs:23]: Moving a compile-time determined value to a runtime call to GetType()?  How un-Toub-like :smile:

File: src/libraries/System.Runtime/tests/System/ObjectDisposedExceptionTests.cs
Comment (bartonjs) [inline src/libraries/System.Runtime/tests/System/ObjectDisposedExceptionTests.cs:50]: We should test that the thrown exception has the ObjectName property set the way we expect.

In this case

```suggestion
            ObjectDisposedException ex = AssertExtensions.Throws<ObjectDisposedException>(
                () => ObjectDisposedException.Throw(obj));
            
            Assert.Equals("System.Object", ex.ObjectName);
```

Then it'd be good to use some variety.

File: src/libraries/System.Runtime/tests/System/ObjectDisposedExceptionTests.cs
Comment (bartonjs) [inline src/libraries/System.Runtime/tests/System/ObjectDisposedExceptionTests.cs:56]: This is an inappropriate use of `var` (it's not receiving the result of a `new` expression or a cast), and a pretty inefficient way of getting a well-known value.

```suggestion
            Type type = typeof(object);
```

**Comment by bartonjs** [comment]
LGTM, but I defer to @stephentoub in case he has last minute second thoughts :smile:

File: src/libraries/Common/tests/System/Net/WebSockets/WebSocketStream.cs
Comment (stephentoub) [inline src/libraries/Common/tests/System/Net/WebSockets/WebSocketStream.cs:180]: For these cases where there's an existing `ThrowIfDisposed()` parameterless method, I would still keep it.  Otherwise, as the changes in this file highlight, we end up duplicating the disposal condition in a bunch of places.

File: src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs
Comment (stephentoub) [inline src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs:90]: Ditto

File: src/libraries/System.ComponentModel.Composition/src/System/ComponentModel/Composition/Hosting/ComposablePartExportProvider.cs
Comment (stephentoub) [inline src/libraries/System.ComponentModel.Composition/src/System/ComponentModel/Composition/Hosting/ComposablePartExportProvider.cs:134]: Ditto.  I'll stop commenting on these.

File: src/libraries/System.Diagnostics.EventLog/src/System/Diagnostics/EventLogInternal.cs
Comment (stephentoub) [inline src/libraries/System.Diagnostics.EventLog/src/System/Diagnostics/EventLogInternal.cs:1049]: Does this build?  I thought this library still built against things other than netcoreapp latest.

**Review by jeffhandley — CHANGES_REQUESTED** [review]

**Review by jeffhandley — COMMENTED** [review]
There are a couple of formatting tweaks needed; otherwise this looks good. I've queued up suggestions for those formatting tweaks; I'll go ahead and commit them.

The test failure is #47968.

**Review by jeffhandley — APPROVED** [review]

**Comment by EgorBo** [comment]
Improvements on ubuntu x64 https://github.com/dotnet/perf-autofiling-issues/issues/2086

**Comment by stephentoub** [comment]
> Improvements on ubuntu x64

Why/how would this change have impacted that particular benchmark?

**Comment by EgorBo** [comment]
> > Improvements on ubuntu x64
> 
> Why/how would this change have impacted that particular benchmark?

this commit was the only suspect in improvements of that benchmark
image

but most likely it's just a noise. I assumed ThrowHelpers helped some IPAddress-related functions to become inlining-friendly

**Pull request merged**
